File: ruby/opalcommand/lib/opalcommand/core/policy.rb
# frozen_string_literal: true

module OpalCommand
  module Core
    module Policy
      module_function

      ORDER = %w[normal watch restricted halted].freeze

      
      DEESCALATION_THRESHOLDS = { 'halted' => 10, 'restricted' => 7, 'watch' => 4 }.freeze 

      METADATA = {
        'normal'     => { description: 'Standard operations', max_retries: 5 },
        'watch'      => { description: 'Elevated monitoring', max_retries: 3 },
        'restricted' => { description: 'Limited operations', max_retries: 2 },
        'halted'     => { description: 'All operations suspended', max_retries: 0 }
      }.freeze

      
      def next_policy(current, failure_burst)
        idx = ORDER.index(current) || 0
        return ORDER[idx] if failure_burst <= 2 

        ORDER[[idx + 1, ORDER.length - 1].min]
      end

      def previous_policy(current)
        idx = ORDER.index(current) || 0
        ORDER[[idx - 1, 0].max]
      end

      
      def should_deescalate?(current, success_streak)
        threshold = DEESCALATION_THRESHOLDS[current]
        return false if threshold.nil?

        success_streak >= threshold 
      end

      def all_policies
        ORDER.dup
      end

      def policy_index(name)
        ORDER.index(name) || -1
      end

      def get_metadata(name)
        METADATA[name]
      end

      
      def check_sla_compliance(elapsed_minutes, sla_minutes)
        return :breached  if elapsed_minutes > sla_minutes
        return :at_risk   if elapsed_minutes > sla_minutes * 0.8 

        :compliant
      end

      
      def sla_percentage(elapsed_minutes, sla_minutes)
        return 0.0 if sla_minutes <= 0

        pct = (elapsed_minutes.to_f / sla_minutes) * 100.0
        pct.round(2) 
      end
    end

    PolicyChange = Struct.new(:from_policy, :to_policy, :reason, :timestamp, keyword_init: true)

    class PolicyEngine
      def initialize(initial: 'normal')
        @mutex   = Mutex.new
        @current = initial
        @history = []
      end

      def current
        @mutex.synchronize { @current }
      end

      
      def escalate(failure_burst, reason: nil)
        @mutex.synchronize do
          prev = @current
          @current = Policy.next_policy(@current, failure_burst)
          
          @history << PolicyChange.new(
            from_policy: prev,
            to_policy: @current,
            reason: reason || 'escalation',
            timestamp: Time.now.to_i
          )
          @current
        end
      end

      def deescalate(success_streak, reason: nil)
        @mutex.synchronize do
          return @current unless Policy.should_deescalate?(@current, success_streak)

          prev = @current
          @current = Policy.previous_policy(@current)
          @history << PolicyChange.new(
            from_policy: prev,
            to_policy: @current,
            reason: reason || 'deescalation',
            timestamp: Time.now.to_i
          )
          @current
        end
      end

      def history
        @mutex.synchronize { @history.dup }
      end

      def reset
        @mutex.synchronize do
          @current = 'normal'
          @history.clear
        end
      end

      def auto_adjust(failure_burst, success_streak)
        @mutex.synchronize do
          prev = @current
          if failure_burst > 2
            @current = Policy.next_policy(@current, failure_burst)
          end
          if success_streak >= 4
            threshold = Policy::DEESCALATION_THRESHOLDS[@current]
            if threshold && success_streak > threshold
              @current = Policy.previous_policy(@current)
            end
          end
          @history << PolicyChange.new(
            from_policy: prev,
            to_policy: @current,
            reason: 'auto_adjust',
            timestamp: Time.now.to_i
          ) if prev != @current
          @current
        end
      end

      def severity_level
        @mutex.synchronize do
          Policy::ORDER.index(@current) || 0
        end
      end
    end

    class PolicyChain
      def initialize(policies)
        @mutex = Mutex.new
        @engines = policies.map { |p| PolicyEngine.new(initial: p) }
      end

      def escalate_all(failure_burst)
        @mutex.synchronize do
          @engines.map { |e| e.escalate(failure_burst) }
        end
      end

      def current_levels
        @mutex.synchronize do
          @engines.map(&:current)
        end
      end

      def most_severe
        @mutex.synchronize do
          @engines.max_by { |e| Policy::ORDER.index(e.current) || 0 }&.current
        end
      end

      def count
        @mutex.synchronize { @engines.length }
      end
    end
  end
end

File: rust/tensorforge/src/policy.rs
use std::sync::Mutex;

const ORDER: [&str; 4] = ["normal", "watch", "restricted", "halted"];

pub fn next_policy(current: &str, failure_burst: usize) -> &'static str {
    let idx = ORDER.iter().position(|v| *v == current).unwrap_or(0);
    if failure_burst <= 2 {
        return ORDER[idx];
    }
    ORDER[usize::min(idx + 1, ORDER.len() - 1)]
}

pub fn previous_policy(current: &str) -> &'static str {
    let idx = ORDER.iter().position(|v| *v == current).unwrap_or(0);
    ORDER[idx.saturating_sub(1)]
}

pub fn should_deescalate(success_streak: usize, current: &str) -> bool {
    let threshold = match current {
        "halted" => 10,
        "restricted" => 7,
        "watch" => 5,
        _ => return false,
    };
    success_streak >= threshold
}

#[derive(Clone, Debug)]
pub struct PolicyMetadata {
    pub name: &'static str,
    pub description: &'static str,
    pub max_retries: usize,
}

const POLICY_METADATA: [PolicyMetadata; 4] = [
    PolicyMetadata {
        name: "normal",
        description: "Standard operations",
        max_retries: 3,
    },
    PolicyMetadata {
        name: "watch",
        description: "Elevated monitoring",
        max_retries: 2,
    },
    PolicyMetadata {
        name: "restricted",
        description: "Limited operations",
        max_retries: 1,
    },
    PolicyMetadata {
        name: "halted",
        description: "All operations suspended",
        max_retries: 0,
    },
];

pub fn get_metadata(policy: &str) -> Option<&'static PolicyMetadata> {
    POLICY_METADATA.iter().find(|m| m.name == policy)
}

pub fn all_policies() -> &'static [&'static str] {
    &ORDER
}

pub fn policy_index(policy: &str) -> Option<usize> {
    ORDER.iter().position(|v| *v == policy)
}

#[derive(Clone, Debug)]
pub struct PolicyChange {
    pub from: String,
    pub to: String,
    pub reason: String,
}

pub struct PolicyEngine {
    current: Mutex<String>,
    history: Mutex<Vec<PolicyChange>>,
}

impl PolicyEngine {
    pub fn new() -> Self {
        Self {
            current: Mutex::new("normal".to_string()),
            history: Mutex::new(Vec::new()),
        }
    }

    pub fn current(&self) -> String {
        self.current.lock().unwrap().clone()
    }

    pub fn escalate(&self, failure_burst: usize) -> String {
        let mut current = self.current.lock().unwrap();
        let next = next_policy(&current, failure_burst);
        if next != current.as_str() {
            let change = PolicyChange {
                from: current.clone(),
                to: next.to_string(),
                reason: format!("escalation: {} failures", failure_burst),
            };
            self.history.lock().unwrap().push(change);
            *current = next.to_string();
        }
        current.clone()
    }

    pub fn deescalate(&self) -> String {
        let mut current = self.current.lock().unwrap();
        let prev = previous_policy(&current);
        if prev != current.as_str() {
            let change = PolicyChange {
                from: current.clone(),
                to: prev.to_string(),
                reason: "deescalation: conditions improved".to_string(),
            };
            self.history.lock().unwrap().push(change);
            *current = prev.to_string();
        }
        current.clone()
    }

    pub fn history(&self) -> Vec<PolicyChange> {
        self.history.lock().unwrap().clone()
    }

    pub fn reset(&self) {
        *self.current.lock().unwrap() = "normal".to_string();
        self.history.lock().unwrap().clear();
    }
}

pub fn check_sla_compliance(actual_minutes: i32, sla_minutes: i32) -> bool {
    actual_minutes <= sla_minutes
}

pub fn sla_percentage(records: &[(i32, i32)]) -> f64 {
    if records.is_empty() {
        return 100.0;
    }
    let compliant = records
        .iter()
        .filter(|(actual, sla)| check_sla_compliance(*actual, *sla))
        .count();
    (compliant as f64 / records.len() as f64) * 100.0
}


pub fn policy_weight(policy: &str) -> f64 {
    match policy {
        "normal" => 4.0,    
        "watch" => 3.0,     
        "restricted" => 2.0, 
        "halted" => 1.0,    
        _ => 0.0,
    }
}


pub fn escalation_needed(failure_rate: f64, threshold: f64) -> bool {
    failure_rate >= threshold  
}


pub fn risk_score(severity: f64, probability: f64) -> f64 {
    severity + probability  
}


pub fn within_grace_period(elapsed_seconds: u64, grace_seconds: u64) -> bool {
    elapsed_seconds < grace_seconds  
}


pub fn max_retries_for_policy(policy: &str) -> usize {
    match policy {
        "normal" => 3,
        "watch" => 2,
        "restricted" => 1,
        "halted" => 0,
        _ => 0,  
    }
}


pub fn cooldown_multiplier(policy: &str) -> u64 {
    match policy {
        "normal" => 1,
        "watch" => 1,
        "restricted" => 1,
        "halted" => 1,
        _ => 1,
    }
}

pub fn cooldown_remaining(last_change_at: u64, now: u64, cooldown_duration: u64) -> u64 {
    let elapsed = now.saturating_sub(last_change_at);
    cooldown_duration.saturating_sub(elapsed)
}

pub fn aggregate_risk(events: &[(f64, f64)]) -> f64 {
    if events.is_empty() { return 0.0; }
    let individual_risks: Vec<f64> = events.iter()
        .map(|&(severity, probability)| severity * probability)
        .collect();
    let max_risk = individual_risks.iter().cloned().fold(0.0_f64, f64::max);
    let sum: f64 = individual_risks.iter().sum();
    let avg_others = (sum - max_risk) / (events.len() as f64).max(1.0);
    (max_risk + avg_others) / 2.0
}

pub fn escalation_with_cooldown(
    current: &str,
    failure_burst: usize,
    _last_change_at: u64,
    now: u64,
    min_cooldown: u64,
) -> &'static str {
    if now < min_cooldown {
        return ORDER[ORDER.iter().position(|v| *v == current).unwrap_or(0)];
    }
    next_policy(current, failure_burst)
}
